python/tvm/relax/transform/legalize_ops/manipulate.py

```diff
@@ -205,3 +213,16 @@ def te_layout_transform(data, name):
     output_dtype = call_args[0].struct_info.dtype
     output_sinfo = [TensorStructInfo(output_shape, output_dtype)]
     return call_tir(gvar, call_args, output_sinfo, tir_vars)
+
+
+@register_legalize("relax.remove_pad")
+def _remove_pad(bb: BlockBuilder, call: Call) -> Expr:
+    orig_shape = call.attrs.orig_shape
+
+    def te_remove_pad(data):
+        """
+        Returns a new compute that restrict the original expression to the shape of orig_shape
+        """
+        return te.compute(orig_shape, data, name="te_remove_pad")
```

Comment:
How do you know that this removes the correct padded values?

Reply:
Hi @masahi, currently the elements are padded in the end of the original buffer. So, when we iterate over orig_shape(dimensions of non-padded buffer), we can be sure that we will be iterating over non-padded entries.
The one assumption that I have taken here is that padding will always be happening in the end of the buffer.
In future, if AlterOp introduces padding in-between original buffer then we should modify remove_pad operator to reflect the same.

Reply:
> currently the elements are padded in the end of the original buffer

Isn't that a specific behavior of the layout transform pass? Here you are defining a general function that undoes padding, and padding can be anything along the begin / end of each axis.

Reply:
Technically the behavior of remove_pad can be extended to remove padding along the given axis and given offset. In this PR, the support is added for AlterOp pass explicitly, to remove the padding which is being introduced by layout transform pass. In the next PR, I plan to work on generalizing remove_pad functionality to be generic.

Will it be good idea to separate out this PR from generalizing remove_pad functionality?

Reply:
We shouldn't introduce a half-baked op definition that is meant to be used only by a certain pass. The proper order is that a generic op definition should come first.

Reply:
Better yet, is this actually needs to be an "op"? Since the op definition is so trivial, how about moving this compute definition to `alter_op_layout.cc` and use it there? You can remove all boilerplate and test cases required for a Relax op definition.

Reply:
For example, https://github.com/mlc-ai/mlc-llm uses weight quantize / dequantize functions (called `encode / decode`), RMS norm etc implemented in TE without introducing Relax ops for them. They are represented by `call_tir`.

Reply:
Thanks @masahi for your helpful suggestions. In the latest changes I have defined remove_pad operator inside `alter_op_impl.cc`.
As of now, the test case in this PR fails because the `te::compute` introduces min Range value to `int32(0)`, whereas when reading `Expected` relax testcase the min Range value is being read as `int64(0)`.
Please let me know what I should do with min Range problem.

Reply:
Can you show the error?

Reply:
Hi @masahi, the error looks something like this,

`ValueError: StructuralEqual check failed, caused by lhs at <root>.functions[I.GlobalVar("remove_pad")].body.block.body.body.block.iter_vars[0].dom.min.dtype and rhs at <root>.functions[I.GlobalVar("remove_pad")].body.block.body.body.block.iter_vars[0].dom.min.dtype`

where the lhs dtype is coming out to be `int32` when running the `alter_op_impl `pass, and rhs dtype is `int64` which is coming from `Expected module`.

I have commited a change in `src/te/operation/compute_op.cc` which is resolving the issue.
My question is, does the change in `src/te/operation/compute_op.cc` should go as a separate PR or can I have it in this PR only?

Reply:
If all tests pass, I think it is fine to include such change in this PR.